### ISSUE TITLE
How to predict the whole validation dataset in Imagenet using pretrained model VGG-16 in KERAS on Theano backend?

### ISSUE THREAD

**Issue by aditbhrgv** [keras-team/keras#9023]
Please make sure that the boxes below are checked before you submit your issue. If your issue is an implementation question, please ask your question on [StackOverflow](http://stackoverflow.com/questions/tagged/keras) or [join the Keras Slack channel](https://keras-slack-autojoin.herokuapp.com/) and ask there instead of filing a GitHub issue.

Thank you!

- [x] Check that you are up-to-date with the master branch of Keras. You can update with:
pip install git+git://github.com/keras-team/keras.git --upgrade --no-deps

- [x] If running on TensorFlow, check that you are up-to-date with the latest version. The installation instructions can be found [here](https://www.tensorflow.org/get_started/os_setup).

- [x] If running on Theano, check that you are up-to-date with the master branch of Theano. You can update with:
pip install git+git://github.com/Theano/Theano.git --upgrade --no-deps

- [ ] Provide a link to a GitHub Gist of a Python script that can reproduce your issue (or just copy the script here if it is short).

Hi, 

I am using pre-trained VGG16 model to do inference on Imagenet dataset. I am able to do correct inference on one image 'cat.jpg' .However, I want to test the whole validation dataset of Imagenet. How should I do this in Keras with Theano backend?

How should I load validation  dataset(50000 images) in theano ? and do model.predict() on whole dataset ?

THanks for your suggestions in advance!

**Comment by tomastheod-ITI** [comment]
@aditbhrgv You could load the data set in small batches, predict the class for each batch of images and append the results in a list. You will end up with a list that holds the predictions for the whole data set. 

Experiment with different batch sizes to see if it affects the model predictions and to adjust the overall run time. model.predict() also takes a batch_size argument.

**Comment by aditbhrgv** [comment]
What are the ground truths used for testing validation set on ILSVRC-2012 dataset ? I got only :
Top1 accuracy
41.412
Top5 accuracy
66.472
I used weights file "vgg16_weights_th_dim_ordering_th_kernels.h5" instead of "vgg16_weights.h5" since it gave compilation errors.

I have tried to mean subtract from images  via dataGenerator.mean

VALIDATION_IMAGES = 50000
datagen = ImageDataGenerator(rescale=1.,featurewise_center=True)
datagen.mean = np.array([103.939, 116.779,123.68,], dtype=np.float32).reshape((3, 1, 1))
generator = datagen.flow_from_directory(
        'data/validation',
        target_size=(224, 224),
        batch_size=1,
        class_mode=None,
        shuffle=False)
...
    ground_truth = np.loadtxt("val.txt")
    print ground_truth[0]
    print ground_truth[1]
    print ground_truth[2]
    print ground_truth[3]

    bottleneck_features_validation = model.predict_generator(generator, VALIDATION_IMAGES)
    for ii in range(VALIDATION_IMAGES):
        print np.argmax(bottleneck_features_validation[ii])
        if  (np.argmax(bottleneck_features_validation[ii]) == ground_truth[ii]):
          top_1 = top_1 + 1
    print (bottleneck_features_validation.shape)

    top_n_preds = np.argpartition(bottleneck_features_validation, -5)[:,-5:]
    print top_n_preds.shape
    print top_n_preds

    for ii in range(VALIDATION_IMAGES):
        if( ground_truth[ii] in top_n_preds[ii]):
           top_5 = top_5 + 1

    print top_1
    print top_5
    print "Top1 accuracy "
    print (top_1/50000.0)*100
    print "Top5 accuracy "
    print (top_5/50000.0)*100



However , I think I have to transpose images into BGR. I don't know how to transpose via imageDatagenerator?


I used ground truth from this link: (val.txt)
https://github.com/BVLC/caffe/blob/master/data/ilsvrc12/get_ilsvrc_aux.sh

How does this Keras model claims to achieve "achieves 7.5% top-5 error on ILSVRC-2012-val" ?
please let me know which ground truth is being used here ..

Thanks

**Comment by aditbhrgv** [comment]
Now, I used the below wrapper as a preprocess_input_vgg function to the arguement of ImageDataGenerator() and I got 

Top1 accuracy
64.722
Top5 accuracy
85.736

on validation dataset of 50000 images: Still wondering it's less than expected..Why ? 


def preprocess_input_vgg(x):
    """Wrapper around keras.applications.vgg16.preprocess_input()
    to make it compatible for use with keras.preprocessing.image.ImageDataGenerator's
    `preprocessing_function` argument.
    
    Parameters
    ----------
    x : a numpy 3darray (a single image to be preprocessed)
    
    Note we cannot pass keras.applications.vgg16.preprocess_input()
    directly to to keras.preprocessing.image.ImageDataGenerator's
    `preprocessing_function` argument because the former expects a
    4D tensor whereas the latter expects a 3D tensor. Hence the
    existence of this wrapper.
    
    Returns a numpy 3darray (the preprocessed image).
    
    """
    from keras.applications.vgg16 import preprocess_input
    X = np.expand_dims(x, axis=0)
    X = preprocess_input(X)
    return X[0]

**Comment by calebrob6** [comment]
@aditbhrgv not sure if this is still relevant to you, and I've also posted on issue 8672, but I've written up the steps I have found for reproducing the top-1/top-5 accuracy reported in the Keras documentation here, http://calebrob.com/ml/imagenet/ilsvrc2012/2018/10/22/imagenet-benchmarking.html. The code is here, https://github.com/calebrob6/imagenet_validation.

Best.
